plutus-wallet-api/ledger/Ledger/Scripts.hs

```diff
@@ -229,6 +267,46 @@ to the previous problem: apply a function which does a pattern match and returns
 otherwise. Then, as before, we just check for error in the overall evaluation.
 -}
 
+{- Note [Sealing data scripts]
+To pass data scripts securely through the redeemer, we need to *seal* them. This means applying the 'seal' builtin
+to the data script value, and then applying the redeemer to all the (sealed) values.
+
+However, 'seal' is *polymorphic*, so we need to instantiate it at the correct type first. So we need to know the
+type of all the data scripts, which means we need to typecheck them. This is annoying, but I don't have
+another solution at the moment.
+-}
+
+{- Note [Impredicative polymorphism and seal]
+We need to turn 'seal :: forall a . a -> Sealed a' into a PLC term, which we
+can instantiate at a PLC type (which we do *not* have the corresponding Haskell type
+for). But this causes GHC to get its knickers in a twist in two ways:
+- GHC is very eager to replace the quantified type variable with 'Any', which we can't
+currently reconstruct. See Note [Polymorphic values and Any] for more complaining.
+- If you try and specify the type you end up asking for a 'CompiledCode (forall a . a -> Sealed a)'.
+But this is impredicative polymorphism, and Banned.
+
+So we apply the usual trick of wrapping it in a newtype wrapper. However, we then need
+to *unwrap* it to actually apply the function, and the function which does *that* is
+also going to be polymorphic, so we hit exactly the same problem defining it.
+
+At this point we deploy a CURSED HACK. We compile newtypes into precisely their
+underlying type... so we actually don't have to unwrap it at all! We just leave
+it there to satisfy GHC and act as though it's unwrapped later. It's horrible, but
+it seems to work.
```

Comment:
Am I a terrible person if I think that even though it's clearly hacky, it's not that horrible?

Reply:
:speak_no_evil: